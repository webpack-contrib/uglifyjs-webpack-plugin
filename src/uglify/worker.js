import minify from './minify';

module.exports = (options, callback) => {
  try {
    // 'use strict' => this === undefined (Clean Scope)
    // Safer for possible security issues, albeit not critical at all here
    // eslint-disable-next-line no-new-func, no-param-reassign
    options = new Function(`'use strict'\nreturn ${options}`)();

    callback(null, minify(options));
  } catch (errors) {
    callback(errors);
  }
};
