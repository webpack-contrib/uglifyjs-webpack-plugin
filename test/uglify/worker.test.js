import serialize from 'serialize-javascript';
import worker from '../../src/uglify/worker';

describe('matches snapshot', () => {
  it('normalizes when options.extractComments is regex', () => {
    const options = {
      file: 'test1.js',
      input: 'var foo = 1;/* hello */',
      extractComments: /foo/,
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when uglifyOptions.output.comments is string: all', () => {
    const options = {
      file: 'test2.js',
      input: 'var foo = 1;/* hello */',
      uglifyOptions: {
        output: {
          comments: 'all',
        },
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when uglifyOptions.output.comments is string: some', () => {
    const options = {
      file: 'test3.js',
      input: 'var foo = 1;/* hello */',
      uglifyOptions: {
        output: {
          comments: 'some',
        },
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when uglifyOptions.extractComments is number', () => {
    const options = {
      file: 'test4.js',
      input: 'var foo = 1;/* hello */',
      uglifyOptions: {
        output: {
          comments: 'some',
        },
      },
      extractComments: 1,
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('when applied with extract option set to a single file', () => {
    const options = {
      file: 'test5.js',
      input: '/******/ function hello(a) {console.log(a)}',
      uglifyOptions: {
        output: {
          comments: 'all',
        },
      },
      extractComments: {
        condition: 'should be extracted',
        filename(file) {
          return file.replace(/(\.\w+)$/, '.license$1');
        },
        banner(licenseFile) {
          return `License information can be found in ${licenseFile}`;
        },
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('when options.inputSourceMap', () => {
    const options = {
      file: 'test6.js',
      input: 'function foo(x) { if (x) { return bar(); not_called1(); } }',
      inputSourceMap: {
        version: 3,
        sources: ['test1.js'],
        names: ['foo', 'x', 'bar', 'not_called1'],
        mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('when uglifyOptions.keep_fnames is enabled', () => {
    const options = {
      file: 'test7.js',
      input: 'function foo(){ return function bar() { return a; }(); }',
      uglifyOptions: {
        keep_fnames: true,
      },
    };

    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
      expect(data.code).toEqual(expect.stringContaining('bar'));
    });
  });

  it('when uglifyOptions.keep_classnames is enabled', () => {
    const options = {
      file: 'test8.js',
      input: 'function foo(){ return class Component { }; }',
      uglifyOptions: {
        keep_classnames: true,
      },
    };

    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
      expect(data.code).toEqual(expect.stringContaining('Component'));
    });
  });

  it('when uglifyOptions.safari10 is enabled', () => {
    const options = {
      file: 'test9.js',
      input: 'function f(o) { for (let n of a) { console.log(n); } }',
      uglifyOptions: {
        safari10: true,
      },
    };

    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });
});
