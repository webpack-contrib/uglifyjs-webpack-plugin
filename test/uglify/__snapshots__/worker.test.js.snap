// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test1.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test2.js 1`] = `
Object {
  "code": "var foo=1;/* hello */",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test3.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test4.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test5.js 1`] = `
Object {
  "code": "/******/function hello(o){console.log(o)}",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test6.js 1`] = `
Object {
  "code": "function foo(f){if(f)return bar()}",
  "error": undefined,
  "extractedComments": Array [],
  "map": "{\\"version\\":3,\\"sources\\":[\\"test1.js\\"],\\"names\\":[\\"foo\\",\\"x\\",\\"not_called1\\"],\\"mappings\\":\\"AAAA,SAASA,IAAIC,GACT,GACIA,EAAA,OACAC\\"}",
  "warnings": undefined,
}
`;

exports[`test7.js 1`] = `
Object {
  "code": "function foo(){return function bar(){return a}()}",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test8.js 1`] = `
Object {
  "code": "function foo(){return class Component{}}",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`test9.js 1`] = `
Object {
  "code": "function f(o){for(let f of a)console.log(f)}",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;
